### ISSUE TITLE
SC4-1-1-accesskey test procedure may only check first accesskey 

### ISSUE THREAD

**Issue by EmmaJP** [auto-wcag/auto-wcag#43]
**Selector**

Select any element with an accesskey attribute. The following CSS selector could be used: *[accesskey]

**Step 1**

- IF there is no list called 'knownAccesskeys':
  - Create an empty list 'knownAccesskeys;
- Take the first character of the ACCESSKEY attribute as FirstChar
- IF the FirstChar exists in 'knownAccesskeys':
  - Return step1-fail
- ELSE:
  - Add the FirstChar to 'knownAccesskeys'
  - Return step1-pass

...

Fail "The accesskey of this element has occurred before on this page"

**Comment by EmmaJP** [comment]
I had suggested the following as steps:

**Step 1**

If there is no list called 'knownAccesskeys', create an empty list 'knownAccesskeys'.

**Step 2**

For each element, check if the first character of the accesskey attribute value exists in 'knownAccesskeys'.

if yes, return [step2-fail](#step2-fail)

else, add the first character of the accesskey attribute value for the selected element to 'knownAccesskeys'

**Step 3**

Confirm there were no fails.

if yes, return [step3-pass](#step3-pass)

...

Fail "Accesskey <attribute-value> is duplicated on the page."

Pass "All accesskey attributes are unique on the page."

**Comment by EmmaJP** [comment]
Based on the comment on https://github.com/auto-wcag/auto-wcag/issues/46 it may be possible to use the node list from the selector as the list of known access keys.

Something like below? With additional assumption and fail/pass as previously suggested.

...

**Selector**

Select all elements that match the following CSS selector: *[accesskey]

**Step 1**

For each element, check if the first character of the accesskey attribute matches the first character of any other selected element's accesskey attribute.

if yes, return step1-fail

**Step 2**

Confirm there were no fails.

if yes, return step2-pass

...

**Comment by WilcoFiers** [comment]
I suggest:

**Step 1**

Check if the first character of the accesskey attribute matches the first character of any other selected element's accesskey attribute.

if yes, return step1-fail

Else return step2-pass

**Comment by EmmaJP** [comment]
As with unique ids, it should check the whole page before returning a pass. The success criteria requires all accesskeys in the web-page to be unique, so it cannot pass just because one accesskey is unique. Hence the additional assumption as well as the second step before returning a pass.

**Comment by 2biazdk** [comment]
If one fails, it never reaches step 2, does it?

**Comment by EmmaJP** [comment]
True, but if the first element passes that doesn't make the whole page pass. So this does need a step to loop through all the elements before a pass can be returned.

I guess it is up to the person making the test tool to decide if it returns only a fail as soon as it hits one, or to collect up all fails for a document and return them as a set, so a developer can more readily fix them all at once rather than one at a time.

**Comment by 2biazdk** [comment]
I think the general idea is to flag each possibly infringing element as either passed or failed, and then you can treat that list of elements as you'd like afterwards.

**Comment by EmmaJP** [comment]
Based on above and further comment on https://github.com/auto-wcag/auto-wcag/issues/46 I guess something like this then:

...

**Assumptions**

- Assumes a full page has loaded to enable comparison.
- If the accesskey value has multiple characters the user agent picks the first character and ignores the rest. See: WHATWG on Interaction.

...

**Selector**

Select all elements that match the following CSS selector: *[accesskey]

**Step 1**

For each element, check if the first character of the accesskey attribute matches the first character of any other selected element's accesskey attribute.

if yes, return step1-fail

else, return step1-pass

...

Fail: "Accesskey [attribute-value] is duplicated on the page."
Pass: "Accesskey [attribute-value] is unique on the page."

**Comment by WilcoFiers** [comment]
This looks good. Please create a pull request and then close this issue.
